File: objc/State/State/MouseCursor.m
#import "MouseCursor.h"
#import "AbstractTool.h"

@implementation MouseCursor {
    id<AbstractTool> tool;
}

-(instancetype)init {
    if(self = [super init]) {
        tool = [[PenTool alloc] init];
    }
    return self;
}

-(void)mouseTo:(Point)p {
    [tool mouseTo:p];
}

-(void)mouseDown:(Point)p {
    [tool mouseDown:p];
}

-(void)mouseUp:(Point)p {
    [tool mouseUp:p];
}

-(void)usePenTool {
    tool = [[PenTool alloc] init];
}

-(void)useTriangleTool {
    tool = [[TriangleTool alloc] init];
}


@end

File: swift/State/State/MouseCursor.swift
import Foundation

class MouseCursor {
    var tool: AbstractTool
    
    init() {
        tool = PenTool()
    }
    
    func mouseTo(_ p: Point) -> Void {
        tool.mouseTo(p)
    }
    
    func mouseDown(_ p: Point) -> Void {
        tool.mouseDown(p)
    }
    
    func mouseUp(_ p: Point) -> Void {
        tool.mouseUp(p)
    }
    
    func usePenTool() -> Void {
        tool = PenTool()
    }
    
    func useTriangleTool() -> Void {
        tool = TriangleTool()
    }
    
}
